petastorm/reader_impl/row_group_loader.py

```diff
@@ -0,0 +1,211 @@
+    """Merges two dictionaries together. If the same key is present in both input dictionaries, the value from 'b'
+    dominates."""
+    result = a.copy()
+    result.update(b)
+    return result
+
+
+def _select_cols(a_dict, keys):
+    """Filteres out entries in a dictionary that have a key which is not part of 'keys' argument. `a_dict` is not
+    modified and a new dictionary is returned."""
+    if keys == list(a_dict.keys()):
+        return a_dict
+    else:
+        return {field_name: a_dict[field_name] for field_name in keys}
+
+
+class RowGroupLoader(object):
+    def __init__(self, dataset_url, schema, ngram, local_cache, worker_predicate):
+        """RowGroupLoader responsible for loading one rowgroup at a time. Rows returned are returned encoded.
+
+        :param dataset_url: A url of a parquet dataset.
+        :param schema: A unischema corresponding to the data in the dataset
+        :param ngram: An instance of NGram if ngrams should be read or None, if each row in the dataset corresponds to
+          a single sample returned.
+        :param local_cache: An instance of a rowgroup cache (CacheBase interface) object to be used.
+        :param worker_predicate: An instance of predicate (PredicateBase interface)
+        """
+        self._dataset_url_parsed = urlparse(dataset_url)
+        self._schema = schema
+        self._ngram = ngram
+        self._local_cache = local_cache
+        self._worker_predicate = worker_predicate
+
+        resolver = FilesystemResolver(self._dataset_url_parsed)
+        self._dataset = pq.ParquetDataset(
+            resolver.parsed_dataset_url().path,
+            filesystem=resolver.filesystem(),
+            validate_schema=False)
+
+    def load(self, rowgroup_spec):
+        """Loads data form a single rowgroup from the dataset.
+
+        Reads a single rowgroup from a dataset. Returns a list of dictionary with still encoded data.
+        If worker_predicate was passed to the constructor, the predicate is first applied to the columns specified
+        by the predicate. The rest of the columns are loaded only if at least one row matches the predicate.
+
+        A rowgroup will be loaded from local cache, if cache contains an instance of the rowgroup.
+
+        If ngram not None was passed to the constructor, the function returns a dictionary structured according to
+        NGram definition.
+
+        :param rowgroup_spec: A dictionary containing the following fields: 'row_group': ParquetDatasetPiece object
+          describing a rowgroup to be loaded; 'shuffle_row_drop_partition' a tuple with
+          (this_partition, num_of_partitions)
+        :return: A dictionary indexed by field names, or a dictionary defined by NGram spec.
+        """
+        piece = rowgroup_spec['row_group']
+        shuffle_row_drop_partition = rowgroup_spec['shuffle_row_drop_partition']
+
+        # Create pyarrow file system
+        parquet_file = ParquetFile(self._dataset.fs.open(piece.path))
```

Comment:
I realize now that we never close the opened file

Reply:
I still do not see the file opened with `self._dataset.fs.open(piece.path))` closed. Unless i am missing something